File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue<int> lo; // max heap
    priority_queue<int,vector<int>,greater<int>> hi; // min heap

    MedianFinder() {
        
    }
    
    void addNum(int num) {
        if ((int)lo.size() == 0 || num <= lo.top())
            lo.push(num);
        else
            hi.push(num);

        int loSize = lo.size(), hiSize = hi.size();

        if (loSize < hiSize){
            lo.push(hi.top());
            hi.pop();
        } else if (loSize > hiSize + 1){
            hi.push(lo.top());
            lo.pop();
        }
    }
    
    double findMedian() {
        if (lo.size() > hi.size())
            return (double)lo.top();
        return ((double)lo.top() + (double)hi.top()) / 2;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.lo = []
        self.hi = []

    def addNum(self, num: int) -> None:
        if len(self.lo) == 0 or num <= self.lo[0]:
            heappush_max(self.lo,num)
        else:
            heappush(self.hi,num)
        
        if len(self.lo) > len(self.hi) + 1:
            heappush(self.hi, heappop_max(self.lo))
        elif len(self.lo) < len(self.hi):
            heappush_max(self.lo, heappop(self.hi))


    def findMedian(self) -> float:
        if len(self.lo) > len(self.hi):
            return self.lo[0]
        return (self.lo[0] + self.hi[0]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()